### PR TITLE
Split lib and bin into crates and minor refactoring

### LINKED ISSUE #48: Rework
I find the code a little messy, maybe a rework is due.

This issue is to discuss the api.

### PR DESCRIPTION
**Pull request by IogaMaster**
this will close #48 

## changelog
- merge workflows `pull_request.yml` and `main.yml` into `check_rust.yml`
- remove the `benches/` which were very non-mature
- move *tool* functions to `crates/nufmt/src/lib.rs`
- split code base into `nufmt` and `nufmt_cli` inside `crates/`

## Justification
1. **Improved Readability**: Decoupling the code into modular packages can enhance code readability. It allows developers to focus on one specific aspect of the code at a time, making it easier to understand.

2. **Reduced Cognitive Load**: With decoupled packages, developers don't need to grasp the entire codebase in one go. They can delve into specific modules as needed, reducing the cognitive load and making it more accessible.

3. **Easier Onboarding**: When new team members join nufmt, a decoupled codebase makes it easier for them to get up to speed.

4. **Maintainability**: A more understandable codebase is inherently more maintainable. Developers can make changes and improvements more confidently, reducing the risk of introducing new issues.

5. **Long-term Viability**: A codebase that is hard to understand can become a liability in the long run. By investing in a rewrite now, we can ensure the long-term viability and sustainability of nufmt.

### CODE REVIEW

**Comment by IogaMaster** [comment]
Alright all basic functionality is restored in this pr. Please test and let me know what else is in need of changing.

I still need to work on:
- Setting the CI back up

File: docs/specification.md
Comment (amtoine) [inline docs/specification.md:1]: i would like this file to be resurrected :yum:

File: scripts/update-nushell-version
Comment (amtoine) [inline scripts/update-nushell-version:1]: i would like this file to be resurrected :yum:

File: test.nu
Comment (amtoine) [inline test.nu:1]: this file should stay called `toolkit.nu`, it's not a test file, but rather a pure-Nushell `Makefile` alternative :+1:

File: src/lib.rs [hunk 1]
```diff
@@ -1,133 +0,0 @@
...
-
-/// format a string of Nushell code
-pub fn format_string(input_string: &String, config: &Config) -> String {
-    let contents = input_string.as_bytes();
-    let formatted_bytes = format_inner(contents, config);
-    String::from_utf8(formatted_bytes).unwrap()
-}
-
-#[cfg(test)]
-mod test {
```
Comment (amtoine) [inline src/lib.rs:41]: i do not see these tests anywhere now...

can you put them back in the `nufmt` crate? :innocent:

**Review by amtoine — CHANGES_REQUESTED** [review]
now that i have a justification for this PR, i think the changes are pretty great :yum: 

i have put the details of the changelog in the PR description above :+1: 

however, there are a few points in the review threads that need to be addressed before landing this PR :wink: :relieved: 

@AucaCoyan
if you can try the *new* `nufmt` on some files and run the tests and tell me that looks good, one the review threads are closed, then let's go with these changes :partying_face:

File: docs/specification.md
Comment (IogaMaster) [reply docs/specification.md:1]: From a comment on the issue from Auca, he said there are a lot of different standards being worked on at the same time. Maybe we need and RFC to talk about making a unified language spec?

File: test.nu
Comment (IogaMaster) [reply test.nu:1]: I was using that to test formatting and forgot to rename it. :p

File: src/lib.rs (hunk 1 above)
Comment (IogaMaster) [reply src/lib.rs:41]: I don't know why tests arent running, I will investigate

File: docs/specification.md
Comment (amtoine) [reply docs/specification.md:1]: yeah, that's fair :+1: 

but for the time being, this is the only place we found for this file...
so, until we have a proper spec somewhere, i'd rather leave it there :relieved:

File: test.nu
Comment (amtoine) [reply test.nu:1]: no worries, as long as you rename it, that's fine :wink:

File: src/lib.rs (hunk 1 above)
Comment (amtoine) [reply src/lib.rs:41]: well, if the tests are not in the source base anywhere, they won't run, right :wink:

**Comment by AucaCoyan** [comment]
I would close this as this PR didn't see the light in a while.
Is it okay @fdncred ?

**Comment by fdncred** [comment]
ya, it has been a while

**Pull request closed without merging**